lib/Dialect/BaseModelica/Transforms/CallCSE.cpp

```diff
@@ -9,67 +9,234 @@ namespace mlir::bmodelica {
 using namespace ::mlir::bmodelica;
 
 namespace {
-class CallCSEPass : public impl::CallCSEPassBase<CallCSEPass> {
+class CallCSEPass final : public impl::CallCSEPassBase<CallCSEPass> {
 public:
   using CallCSEPassBase<CallCSEPass>::CallCSEPassBase;
 
   void runOnOperation() override;
 
 private:
   mlir::LogicalResult processModelOp(ModelOp modelOp);
+
+  /// Replace all calls in the equivalence group with gets to a generated
+  /// variable. The variable will be driven by an equation derived from the
+  /// first call in the group.
+  ///
+  /// One variable and driver equation will be emitted per result,
+  /// if the call is to a function with multiple result values.
+  void emitCse(llvm::SmallVectorImpl<CallOp> &equivalenceGroup, ModelOp modelOp,
+               DynamicOp dynamicOp, mlir::SymbolTable &symbolTable,
+               mlir::RewriterBase &rewriter);
 };
-} // namespace
 
-void CallCSEPass::runOnOperation() {
-  llvm::SmallVector<ModelOp, 1> modelOps;
+/// Get all call operations in the model.
+void collectCallOps(ModelOp modelOp, llvm::SmallVectorImpl<CallOp> &callOps) {
+  llvm::SmallVector<EquationInstanceOp> dynamicEquationOps;
+  modelOp.collectMainEquations(dynamicEquationOps);
 
-  walkClasses(getOperation(), [&](mlir::Operation *op) {
-    if (auto modelOp = mlir::dyn_cast<ModelOp>(op)) {
-      modelOps.push_back(modelOp);
+  llvm::SmallVector<EquationTemplateOp> templateOps;
+
+  for (EquationInstanceOp equationOp : dynamicEquationOps) {
+    EquationTemplateOp templateOp = equationOp.getTemplate();
+    if (!templateOp.getInductionVariables().empty() ||
+        llvm::is_contained(templateOps, templateOp)) {
```

Comment:
This is O(n^2). It would be better to check for already visited templates by using a set, while at the same time keeping the vector for deterministic output.

Reply:
Yeah ok. After building up the set I need to order the operations though. Initially I was moving the elements into a vector and calling std::sort on it, but while it seemed to work I was a little unsure about how it ordered the operations. 

Is it alright to call sort directly, or would you add a custom comparator. If so what would you compare them on?

Reply:
Yeah, I'll fix it